plugins/modules/ndo_l3out_interface_group_policy.py

```diff
@@ -0,0 +1,629 @@
+            mso.existing = match.details
+        else:
+            mso.existing = {}  # When the state is absent
+    elif module.check_mode and state != "query":  # When the state is present/absent with check mode
+        mso_template.update_config_with_template_and_references(mso.proposed, reference_dict)
+        mso.existing = mso.proposed if state == "present" else {}
+
+    mso.exit_json()
+
+
+def check_authentication_requirements(mso, protocols_dict):
+    failed_attributes = {}
+
+    for protocol_name, protocol_dict in protocols_dict.items():
+        if protocol_dict is not None:
+            if (
+                protocol_dict.get("enable") is True
+                and (protocol_dict.get("authentication") is True or protocol_dict.get("authentication_type") in ["md5", "simple"])
+                and (protocol_dict.get("key") is None or protocol_dict.get("key_id") is None)
+            ):
+                failed_attributes[protocol_name] = "key and key_id are required under {0}".format(protocol_name)
+            elif (protocol_dict.get("enable") is not True) and (
+                protocol_dict.get("authentication") is not None or protocol_dict.get("key") is not None or protocol_dict.get("key_id") is not None
+            ):
+                failed_attributes[protocol_name] = "{0} must be enabled in order to use authentication, key and key_id".format(protocol_name)
+
+    if failed_attributes:
+        failure_messages = " | ".join(failed_attributes.values())
+        mso.fail_json(msg=failure_messages)
+
+
+def update_mso_values_attributes_payload(mso_values, protocols_dict):
+    for protocol_name, protocol_dict in protocols_dict.items():
+        if protocol_dict is not None:
+            if protocol_dict.get("enable") is not None:
+                mso_values[protocol_name] = {
+                    "enabled": protocol_dict.get("enable"),
+                }
+            if protocol_dict.get("enable") is True:
+                if protocol_dict.get("authentication") is True:
+                    mso_values[protocol_name].update(
+                        {
+                            "authEnabled": protocol_dict.get("authentication"),
+                            "keyID": protocol_dict.get("key_id"),
+                            "key": {"value": protocol_dict.get("key")},
+                        }
+                    )
+                elif protocol_dict.get("authentication_type") in ["md5", "simple"]:
+                    mso_values[protocol_name].update(
+                        {
+                            "authType": protocol_dict.get("authentication_type"),
+                            "keyID": protocol_dict.get("key_id"),
+                            "key": {"value": protocol_dict.get("key")},
+                        }
+                    )
+
+
+def process_policy(mso_values, mso_template, tenant_template, policy, policy_ref, policy_type):
```

Comment:
is this a generic function that could be reused in other modules? 

should we rename it `process_tenant_policy` since it is specific for tenant templates?